crates/sui-framework/packages/sui-framework/sources/random.move

```diff
@@ -1,37 +1,88 @@
 // Copyright (c) Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
 
-#[allow(unused_use)]
-// This module provides functionality for generating and using secure randomness.
-//
-// Randomness is currently write-only, until user-facing API is implemented.
+/// This module provides functionality for generating secure randomness.
+///
+/// Randomness works in rounds. Each round is associated with a monotonically increasing round number and a random value
+/// generated by the Sui network.
+///
+/// Generating randomness for programs is a 2-step process:
+/// 1. Create a RandomGeneratorRequest object that commits to a future round (i.e., one for which the randomness is not
+///    available yet).
+/// 2. When the round is available, the program can convert the request to a RandomGenerator object which is a
+///    secure pseudo-random generator. RandomGenerator can be used to generate random values of various types.
+///    The second step may be completed within 1 epoch after the round is available. Programs should gracefully handle
+///    the case where the second step is not completed within 1 epoch after the round is available.
+///
+/// Events are emitted when new randomness is available, allowing dapps to know when to execute the second step.
+///
+///
+// TODO: update the above description based on the meaning of max_size.
 module sui::random {
+    use std::bcs;
     use std::vector;
+    use sui::event;
+    use sui::dynamic_field;
+    use sui::address::to_bytes;
+    use sui::hmac::hmac_sha3_256;
     use sui::object::{Self, UID};
     use sui::transfer;
-    use sui::tx_context::{Self, TxContext};
+    use sui::tx_context::{Self, TxContext, fresh_object_address};
     use sui::versioned::{Self, Versioned};
 
     // Sender is not @0x0 the system address.
     const ENotSystemAddress: u64 = 0;
-    const EWrongInnerVersion: u64 = 1;
+    const EWrongRoundsVersion: u64 = 1;
     const EInvalidRandomnessUpdate: u64 = 2;
+    const EInvalidRange: u64 = 3;
+    const ERoundNotAvailableYet: u64 = 4;
+    const ERoundTooOld: u64 = 5;
 
     const CURRENT_VERSION: u64 = 1;
+    const RAND_OUTPUT_LEN: u16 = 32;
+
+    // TODO: changes needed on the Sui side:
+    // 1. Rounds are increasing monotonically also between epochs (may also be done here if needed).
+    // 2. Guarantee that the last sent partial sig is committed after EoP (and not to generate more).
+    // 3. First round is 1 (may also be handled here if needed).
+    // 4. len(random value) = 16 bytes (to reduce storage space).
+
+    // TODO: On max_size and storage space:
+    // 1. Should work with the granularity of epochs? or approximation of epochs is enough (e.g., max_size = expected
```

Comment:
I lean towards:
- using num of rounds that is high enough to be above the required number of epochs
- using ref counters - store the last few hours with ref-counter, and those that are non-zero for 2 weeks